### PR TITLE
add GRU export support

### PR DESCRIPTION
**Pull request by anderspapitto**

### CODE REVIEW

File: onnx_caffe2/backend.py [hunk 1]
```diff
@@ -304,39 +305,41 @@ def _create_gemm(cls, init_model, pred_model, n, opset_version):
...
-        assert init_model is not None, "cannot convert LSTMs through this codepath"
-        assert pred_model is not None, "cannot convert LSTMs through this codepath"
-
-        attrs = dict(n.attrs)
-        hidden_size = attrs.pop('hidden_size')
-        assert not attrs, "unsupported LSTM attributes: " + attrs.keys()
-
-        input_blob, W, R, B, sequence_lens, initial_h, initial_c = n.inputs
-
+    def _shape_inference(cls, init_model, pred_model, n, input_blob, W):
```
Comment (dzhulgakov) [inline onnx_caffe2/backend.py]: maybe call it _rnn_shape_inference? it's not too generic

File: onnx_caffe2/backend.py [hunk 2]
```diff
@@ -304,39 +305,41 @@ def _create_gemm(cls, init_model, pred_model, n, opset_version):
...
-                input_size = x.type.tensor_type.shape.dim[2].dim_value
+                return x.type.tensor_type.shape.dim[2].dim_value
         for x in map(OnnxNode, pred_model.graph.node):
             if x.op_type == n.op_type and x.outputs[0] == input_blob:
-                input_size = x.attrs['hidden_size']
+                return x.attrs['hidden_size']
 
+    @classmethod
+    def _create_lstm(cls, init_model, pred_model, n, opset_version):
+        assert init_model is not None, "cannot convert LSTMs through this codepath"
```
Comment (ezyang) [inline onnx_caffe2/backend.py]: I'm not sure what this assert failure means

File: onnx_caffe2/backend.py [hunk 3]
```diff
@@ -384,6 +387,70 @@ def _create_lstm(cls, init_model, pred_model, n, opset_version):
                          list(init_net.Proto().op),
                          list(pred_mh.Proto().external_input))
 
+    @classmethod
+    def _create_gru(cls, init_model, pred_model, n, opset_version):
+        assert init_model is not None, "cannot convert GRUs through this codepath"
+        assert pred_model is not None, "cannot convert GRUs through this codepath"
+
+        attrs = dict(n.attrs)
```
Comment (ezyang) [inline onnx_caffe2/backend.py]: Is this... so that you make a copy of the dictionary?

File: onnx_caffe2/backend.py [hunk 4]
```diff
@@ -384,6 +387,70 @@ def _create_lstm(cls, init_model, pred_model, n, opset_version):
...
+        hidden_size = attrs.pop('hidden_size')
+        linear_before_reset = attrs.pop('linear_before_reset')
+        assert not attrs, "unsupported GRU attributes: " + str(attrs.keys())
+
+        input_blob, W, R, B, sequence_lens, initial_h = n.inputs
+
+        input_size = cls._rnn_shape_inference(init_model, pred_model, n, input_blob, W)
+        assert input_size is not None, "best-effort shape inference for GRU input failed"
```
Comment (ezyang) [inline onnx_caffe2/backend.py]: Python asserts can be compiled away, so it's best to do an explicit test and raise exception if user input could induce the error.

**Pull request merged**